sns/parity_math.c

```diff
@@ -194,254 +419,1235 @@ static void (*fidx_recover[M0_PARITY_CAL_ALGO_NR])(struct m0_parity_math *math,
+	return (rc == 0) ? M0_RC(rc) : M0_ERR(rc);
+}
 
-	return 0;
+M0_INTERNAL void m0_parity_math_fail_index_recover(struct m0_parity_math *math,
+						   struct m0_buf *data,
+						   struct m0_buf *parity,
+						   const uint32_t fidx)
+{
+	M0_ENTRY();
+	(*fidx_recover[math->pmi_parity_algo])(math, data, parity, fidx);
+	M0_LEAVE();
 }
 
-M0_INTERNAL void m0_parity_math_fini(struct m0_parity_math *math)
+M0_INTERNAL void m0_parity_math_refine(struct m0_parity_math *math,
+				       struct m0_buf *data,
+				       struct m0_buf *parity,
+				       uint32_t data_ind_changed)
 {
-	if (math->pmi_parity_algo == M0_PARITY_CAL_ALGO_REED_SOLOMON) {
-		vandmat_fini(&math->pmi_vandmat);
-		m0_matrix_fini(&math->pmi_vandmat_parity_slice);
-		m0_matvec_fini(&math->pmi_data);
-		m0_matvec_fini(&math->pmi_parity);
-
-		m0_linsys_fini(&math->pmi_sys);
-		m0_matrix_fini(&math->pmi_sys_mat);
-		m0_matvec_fini(&math->pmi_sys_vec);
-		m0_matvec_fini(&math->pmi_sys_res);
-//		m0_parity_fini();
-	}
+	/* for simplicity: */
+	m0_parity_math_calculate(math, data, parity);
 }
 
-M0_INTERNAL int m0_parity_math_init(struct m0_parity_math *math,
-				    uint32_t data_count, uint32_t parity_count)
+#if !ISAL_ENCODE_ENABLED
+M0_INTERNAL int m0_parity_recov_mat_gen(struct m0_parity_math *math,
+					uint8_t *fail)
 {
-	int ret;
+	int rc;
 
-	M0_PRE(data_count >= 1);
-	M0_PRE(parity_count >= 1);
-	M0_PRE(data_count >= parity_count);
-	M0_PRE(data_count <= SNS_PARITY_MATH_DATA_BLOCKS_MAX);
+	recovery_mat_fill(math, fail,
+			  math->pmi_data_count + math->pmi_parity_count,
+			  &math->pmi_sys_mat);
+	m0_matrix_init(&math->pmi_recov_mat, math->pmi_sys_mat.m_width,
+		       math->pmi_sys_mat.m_height);
+	rc = m0_matrix_invert(&math->pmi_sys_mat, &math->pmi_recov_mat);
 
-	M0_SET0(math);
+	return rc == 0 ? M0_RC(0) : M0_ERR(rc);
+}
 
-	math->pmi_data_count	= data_count;
-	math->pmi_parity_count	= parity_count;
+M0_INTERNAL void m0_parity_recov_mat_destroy(struct m0_parity_math *math)
+{
+	m0_matrix_fini(&math->pmi_recov_mat);
+}
+#endif /* !ISAL_ENCODE_ENABLED */
 
-	if (parity_count == 1) {
-		math->pmi_parity_algo = M0_PARITY_CAL_ALGO_XOR;
-		return 0;
-	} else {
-		math->pmi_parity_algo = M0_PARITY_CAL_ALGO_REED_SOLOMON;
+M0_INTERNAL void m0_parity_math_buffer_xor(struct m0_buf *dest,
+					   const struct m0_buf *src)
+{
+        uint32_t ei; /* block element index. */
+
+        for (ei = 0; ei < src[0].b_nob; ++ei)
+		((uint8_t*)dest[0].b_addr)[ei] ^=
+			(m0_parity_elem_t)((uint8_t*)src[0].b_addr)[ei];
+}
+
+M0_INTERNAL int m0_sns_ir_init(const struct m0_parity_math *math,
+			       uint32_t local_nr, struct m0_sns_ir *ir)
+{
+	int ret;
+
+	M0_ENTRY("math=%p, local_nr=%u, ir=%p", math, local_nr, ir);
+
+	M0_PRE(math != NULL);
+	M0_PRE(ir != NULL);
+
+	M0_SET0(ir);
+
+	ir->si_data_nr		   = math->pmi_data_count;
+	ir->si_parity_nr	   = math->pmi_parity_count;
+	ir->si_local_nr		   = local_nr;
+	ir->si_alive_nr 	   = ir_blocks_count(ir);
```

Comment:
No need to have so many spaces before the assignment operator.

Reply:
Extra spaces removed.